http/obmc_ssh.hpp

```diff
@@ -0,0 +1,221 @@
+#pragma once
+
+#include <app.hpp>
+#include <boost/beast/core/flat_static_buffer.hpp>
+#include <boost/process.hpp>
+#include <websocket.hpp>
+
+#include <csignal>
+
+namespace crow
+{
+
+namespace obmc_ssh
+{
+
+static crow::websocket::Connection* session = nullptr;
+
+static std::array<char, 4096> outputBuffer;
+static std::string inputBuffer;
+
+class Handler : public std::enable_shared_from_this<Handler>
+{
+  public:
+    Handler(boost::asio::io_context& ios) :
+        pipeOut(ios), pipeIn(ios), doingWrite(false)
+    {}
+
+    ~Handler() = default;
+
+    void doClose()
+    {
+        // boost::process::child::terminate uses SIGKILL, need to send SIGTERM
+        int rc = kill(processSSH.id(), SIGTERM);
+        if (rc)
+        {
+            return;
+        }
+        processSSH.wait();
+    }
+
+    void connect()
+    {
+        std::error_code ec;
+        processSSH =
+            boost::process::child("/bin/sh", boost::process::std_out > pipeOut,
+                                  boost::process::std_in < pipeIn, ec);
+        if (ec)
+        {
+            BMCWEB_LOG_ERROR << "Couldn't connect to processSSH: "
+                             << ec.message();
+            if (session != nullptr)
+            {
+                session->close("Error connecting to processSSH");
+            }
+            return;
+        }
+        doWrite();
+        doRead();
+    }
+
+    void doWrite()
+    {
+        if (doingWrite)
+        {
+            BMCWEB_LOG_DEBUG << "Already writing.  Bailing out";
+            return;
+        }
+
+        if (inputBuffer.empty())
+        {
+            BMCWEB_LOG_DEBUG << "inputBuffer empty.  Bailing out";
+            return;
+        }
+
+        doingWrite = true;
+        pipeIn.async_write_some(
+            boost::asio::buffer(inputBuffer.data(), inputBuffer.size()),
+            [this, self(shared_from_this())](boost::beast::error_code ec,
+                                             std::size_t bytesWritten) {
+                BMCWEB_LOG_DEBUG << "Wrote " << bytesWritten << "bytes";
+                doingWrite = false;
+                inputBuffer.erase(0, bytesWritten);
+
+                if (session == nullptr)
+                {
+                    return;
+                }
+                if (ec == boost::asio::error::eof)
+                {
+                    session->close("ssh socket port closed");
+                    return;
+                }
+                if (ec)
+                {
+                    session->close("Error in writing to processSSH port");
+                    BMCWEB_LOG_ERROR << "Error in ssh socket write " << ec;
+                    return;
+                }
+                doWrite();
+            });
+    }
+
+    void doRead()
+    {
```

Comment:
If the client passes "rm -rf /," should we put any thought into stopping that?
Excellent question, but it looks like only the manager can access this, so I think it can perform the rm command.

But still, I am not sure other folks need to answer this question.

Reply:
Do not filter commands.